src/tbb/threading_control.cpp

```diff
@@ -0,0 +1,397 @@
+}
+
+void threading_control_impl::set_active_num_workers(unsigned soft_limit) {
+    __TBB_ASSERT(soft_limit <= my_thread_dispatcher->my_num_workers_hard_limit, nullptr);
+    my_thread_request_serializer->set_active_num_workers(soft_limit);
+    my_permit_manager->set_active_num_workers(soft_limit);
+}
+
+threading_control_impl::threading_control_client threading_control_impl::create_client(arena& a) {
+    pm_client* pm_client = my_permit_manager->create_client(a);
+    thread_dispatcher_client* td_client = my_thread_dispatcher->create_client(a);
+
+    return {pm_client, td_client};
+}
+
+threading_control_impl::client_deleter threading_control_impl::prepare_client_destruction(threading_control_impl::threading_control_client client) {
+    thread_dispatcher_client* td_client = client.second;
+    return {td_client->get_aba_epoch(), td_client->priority_level(), td_client, static_cast<pm_client*>(client.first)};
+}
+
+bool threading_control_impl::try_destroy_client(threading_control_impl::client_deleter deleter) {
+    if (my_thread_dispatcher->try_unregister_client(deleter.my_td_client, deleter.aba_epoch, deleter.priority_level)) {
+        my_permit_manager->destroy_client(*deleter.my_pm_client);
+        return true;
+    }
+    return false;
+}
+
+void threading_control_impl::publish_client(threading_control_impl::threading_control_client client) {
+    my_permit_manager->register_client(static_cast<pm_client*>(client.first));
+    my_thread_dispatcher->register_client(client.second);
+}
+
+void threading_control_impl::register_thread(thread_data& td) {
+    my_cancellation_disseminator->register_thread(td);
+}
+void threading_control_impl::unregister_thread(thread_data& td) {
+    my_cancellation_disseminator->unregister_thread(td);
+}
+
+void threading_control_impl::propagate_task_group_state(std::atomic<uint32_t> d1::task_group_context::*mptr_state,
+                                d1::task_group_context& src, uint32_t new_state)
+{
+    my_cancellation_disseminator->propagate_task_group_state(mptr_state, src, new_state);
+}
+
+std::size_t threading_control_impl::worker_stack_size() {
+    return my_thread_dispatcher->worker_stack_size();
+}
+
+unsigned threading_control_impl::max_num_workers() {
+    return my_thread_dispatcher->my_num_workers_hard_limit;
+}
+
+bool threading_control_impl::check_client_priority(threading_control_impl::threading_control_client client) {
+    return static_cast<pm_client*>(client.first)->is_top_priority();
+}
+
+void threading_control_impl::adjust_demand(threading_control_impl::threading_control_client tc_client, int mandatory_delta, int workers_delta) {
+    pm_client& c = *static_cast<pm_client*>(tc_client.first);
+    my_thread_request_serializer->register_mandatory_request(mandatory_delta);
+    my_permit_manager->adjust_demand(c, mandatory_delta, workers_delta);
+}
+
+thread_control_monitor& threading_control_impl::get_waiting_threads_monitor() {
+    return *my_waiting_threads_monitor;
+}
```

Comment:
What is the reason to have a getter of the field? Is it supposed to do something more intelligent in the future? Why not put it into `public` section instead?